jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp

```diff
@@ -0,0 +1,303 @@
+{
+  JLM_ASSERT(MallocMap_.count(&mallocNode) == 0);
+  return MallocMap_[&mallocNode] = AddPointerObject(PointerObjectKind::MallocMemoryObject);
+}
+
+PointerObject::Index
+PointerObjectSet::CreateGlobalMemoryObject(const delta::node & deltaNode)
+{
+  JLM_ASSERT(GlobalMap_.count(&deltaNode) == 0);
+  return GlobalMap_[&deltaNode] = AddPointerObject(PointerObjectKind::GlobalMemoryObject);
+}
+
+PointerObject::Index
+PointerObjectSet::CreateFunctionMemoryObject(const lambda::node & lambdaNode)
+{
+  JLM_ASSERT(FunctionMap_.count(&lambdaNode) == 0);
+  return FunctionMap_[&lambdaNode] = AddPointerObject(PointerObjectKind::FunctionMemoryObject);
+}
+
+PointerObject::Index
+PointerObjectSet::CreateImportMemoryObject(const rvsdg::argument & importNode)
+{
+  JLM_ASSERT(ImportMap_.count(&importNode) == 0);
+  return ImportMap_[&importNode] = AddPointerObject(PointerObjectKind::ImportMemoryObject);
+}
+
+const std::unordered_map<const rvsdg::output *, PointerObject::Index>&
+PointerObjectSet::GetRegisterMap() const noexcept
+{
+  return RegisterMap_;
+}
+
+const std::unordered_map<const rvsdg::node *, PointerObject::Index>&
+PointerObjectSet::GetAllocaMap() const noexcept
+{
+  return AllocaMap_;
+}
+
+const std::unordered_map<const rvsdg::node *, PointerObject::Index>&
+PointerObjectSet::GetMallocMap() const noexcept
+{
+  return MallocMap_;
+}
+
+const std::unordered_map<const delta::node *, PointerObject::Index>&
+PointerObjectSet::GetGlobalMap() const noexcept
+{
+  return GlobalMap_;
+}
+
+const std::unordered_map<const lambda::node *, PointerObject::Index>&
+PointerObjectSet::GetFunctionMap() const noexcept
+{
+  return FunctionMap_;
+}
+
+const std::unordered_map<const rvsdg::argument *, PointerObject::Index>&
+PointerObjectSet::GetImportMap() const noexcept
+{
+  return ImportMap_;
+}
+
+PointerObject::Index
+PointerObjectSet::NumPointerObjects() const noexcept
+{
+  return PointerObjects_.size();
+}
+
+PointerObject &
+PointerObjectSet::GetPointerObject(PointerObject::Index index)
+{
+  JLM_ASSERT(index <= NumPointerObjects());
+  return PointerObjects_[index];
+}
+
+const PointerObject &
+PointerObjectSet::GetPointerObject(PointerObject::Index index) const
+{
+  JLM_ASSERT(index <= NumPointerObjects());
+  return PointerObjects_[index];
+}
+
+const std::unordered_set<PointerObject::Index> &
+PointerObjectSet::GetPointsToSet(PointerObject::Index idx) const
+{
+  JLM_ASSERT(idx <= NumPointerObjects());
+  return PointsToSets_[idx];
+}
+
+// Makes pointee a member of P(pointer)
+bool
+PointerObjectSet::AddToPointsToSet(PointerObject::Index pointer, PointerObject::Index pointee)
+{
+  JLM_ASSERT(pointer < NumPointerObjects());
+  JLM_ASSERT(pointee < NumPointerObjects());
+  // Registers can not be pointed to
+  JLM_ASSERT(GetPointerObject(pointee).GetKind() != PointerObjectKind::Register);
+
+  auto sizeBefore = PointsToSets_[pointer].size();
+  PointsToSets_[pointer].insert(pointee);
+
+  return PointsToSets_[pointer].size() != sizeBefore;
```

Comment:
Does the[ insert() method from std::unordered_set ](https://en.cppreference.com/w/cpp/container/unordered_set/insert) not already give you that information?

Could you not replace this code with:

```
return PointsToSets_[pioniter].insert(pointee).second
```

I am not saying it is more readable....

Maybe you want to consider to use [HashSet](https://github.com/phate/jlm/blob/master/jlm/util/HashSet.hpp)?

Reply:
`std::unordered_set::insert()` absolutely does, I will use that.

I can use HashSet, I didn't know in what ways it differed from unordered_set, so I went with what I already (kind of) knew.